src/facade/dragonfly_listener.cc

```diff
@@ -174,43 +172,113 @@ error_code Listener::ConfigureServerSocket(int fd) {
   return error_code{};
 }
 
+void Listener::PreAcceptLoop(util::ProactorBase* pb) {
+  per_thread_.resize(pool()->size());
+}
+
 void Listener::PreShutdown() {
 }
 
 void Listener::PostShutdown() {
 }
 
+void Listener::OnConnectionStart(util::Connection* conn) {
+  unsigned id = conn->socket()->proactor()->GetIndex();
+  DCHECK_LT(id, per_thread_.size());
+
+  facade::Connection* facade_conn = static_cast<facade::Connection*>(conn);
+  VLOG(1) << "Opening connection " << facade_conn->GetClientId();
+
+  unique_lock lock{mutex_};
+  int32_t prev_cnt = per_thread_[id].num_connections++;
+  ++conn_cnt_;
+
+  if (id == min_cnt_thread_id_) {
+    DCHECK_EQ(min_cnt_, prev_cnt);
+    ++min_cnt_;
+    for (unsigned i = 0; i < per_thread_.size(); ++i) {
+      auto cnt = per_thread_[i].num_connections;
+      if (cnt < min_cnt_) {
+        min_cnt_ = cnt;
+        min_cnt_thread_id_ = i;
+        break;
+      }
+    }
+  }
+}
+
+void Listener::OnConnectionClose(util::Connection* conn) {
+  // TODO: We do not account for connections migrated to other threads. This is a rare case.
+  unsigned id = conn->socket()->proactor()->GetIndex();
+  DCHECK_LT(id, per_thread_.size());
+  auto& pth = per_thread_[id];
+
+  facade::Connection* facade_conn = static_cast<facade::Connection*>(conn);
+  VLOG(1) << "Closing connection " << facade_conn->GetClientId();
+
+  unique_lock lock{mutex_};
+  int32_t cur_cnt = --pth.num_connections;
+  --conn_cnt_;
+
+  auto mincnt = min_cnt_;
+  if (mincnt > cur_cnt) {
+    min_cnt_ = cur_cnt;
+    min_cnt_thread_id_ = id;
+    return;
+  }
```

Comment:
I'm really surprised you use a regular mutex and not a spin lock for all those small ops